tiledb/sm/serialization/array_schema.cc

```diff
@@ -117,57 +120,62 @@ Status filter_pipeline_to_capnp(
   return Status::Ok();
 }
 
-Status filter_pipeline_from_capnp(
-    const capnp::FilterPipeline::Reader& filter_pipeline_reader,
-    tdb_unique_ptr<FilterPipeline>* filter_pipeline) {
-  filter_pipeline->reset(tdb_new(FilterPipeline));
+static tuple<Status, optional<shared_ptr<Filter>>> filter_constructor(
+    capnp::Filter::Reader reader) {
+  FilterType type = FilterType::FILTER_NONE;
+  RETURN_NOT_OK_TUPLE(
+      filter_type_enum(reader.getType().cStr(), &type), nullopt);
+
+  switch (type) {
+    case FilterType::FILTER_BIT_WIDTH_REDUCTION: {
+      auto data = reader.getData();
+      uint32_t window = data.getUint32();
+      return {
+          Status::Ok(),
+          tiledb::common::make_shared<BitWidthReductionFilter>(HERE(), window)};
+    }
+    case FilterType::FILTER_POSITIVE_DELTA: {
+      auto data = reader.getData();
+      uint32_t window = data.getUint32();
+      return {Status::Ok(),
+              tiledb::common::make_shared<PositiveDeltaFilter>(HERE(), window)};
+    }
+    case FilterType::FILTER_GZIP:
+    case FilterType::FILTER_ZSTD:
+    case FilterType::FILTER_LZ4:
+    case FilterType::FILTER_RLE:
+    case FilterType::FILTER_BZIP2:
+    case FilterType::FILTER_DOUBLE_DELTA: {
+      auto data = reader.getData();
+      int32_t level = data.getInt32();
+      return {
+          Status::Ok(),
+          tiledb::common::make_shared<CompressionFilter>(HERE(), type, level)};
+    }
+    default: {
+      assert(false);
```

Comment:
If we don't expect this code path to ever be hit then this should be an exception. Not finding the a matching filter indicates to me either (1) a version skew or (2) data corruption, so I would lean toward throwing an exception. @Shelnutt2 is there any case where we should not throw?

Reply:
I put the assert(false) because I saw it in FilterCreate::deserialize, and therefore thought it was ok to place there.

Reply:
@ihnorton no I'm good to through here

Reply:
@abigalekim then please throw here. For new code (and refactoring) we should replace the use of `assert(false)` with exceptions in cases like this where the codepath should never be hit in normal circumstances.

Reply:
What error should I use to throw with?

Reply:
`throw std::logic_error`. The message should indicate that invalid data was received from the capnp reader; in particular, it should be crystal clear that it's not some other kind of deserialization.

Reply:
Fixed in what I just pushed